### PR TITLE
Gateway API v0.5.0 API Review

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
This PR represents the diff between v0.4.x and a proposed v0.5.0. This was accomplished by copying the apis/v1alpha2 from master to a branch based on the release-0.4 branch. 

### Current status
We believe GatewayClass, Gateway, and HTTPRoute have met the criteria we defined earlier for graduation to beta ([details](https://docs.google.com/document/d/1F3dcs_J8Y6GVlfMGjRtom5ljovnKa_MMIC34GJxUdBs/edit)). The following work is still in progress: 

* Adding [new conditions to HTTPRoute status](https://github.com/kubernetes-sigs/gateway-api/issues/1077#issuecomment-1081237680)
* A variety of [documentation updates and webhook deployment cleanup](https://github.com/kubernetes-sigs/gateway-api/milestone/5)

### Changes from v0.4
The proposed API is largely identical to what we released in v1alpha2, with the following exceptions:

1. We want to graduate GatewayClass, Gateway, and HTTPRoute to v1beta1 ([More Details](https://docs.google.com/document/d/1F3dcs_J8Y6GVlfMGjRtom5ljovnKa_MMIC34GJxUdBs/edit))
2. We have introduced a new "release channels" concept. The new "experimental" release channel allows us to add new features to Gateway API in a similar way to how we use alpha feature gates in upstream to guard new fields initially. ([More Details](https://deploy-preview-1066--kubernetes-sigs-gateway-api.netlify.app/concepts/versioning/#release-channels-experimental-stable))

### New experimental features
There are 3 new experimental features included in this change:

- Path redirects and rewrites [GEP 726](https://gateway-api.sigs.k8s.io/geps/gep-726/)
- Address matching on TCPRoute and UDPRoute [GEP 735](https://gateway-api.sigs.k8s.io/geps/gep-735/)
- Port matching in ParentRef [GEP 957](https://gateway-api.sigs.k8s.io/geps/gep-957/)

### Future port changes
As a follow up to the previous review, there has been a lot of discussion around Ports, and potential for port ranges, all ports, and auto assignment:

- [All Ports and Port Ranges](https://github.com/kubernetes-sigs/gateway-api/issues/818)
- [Optional Listener Port](https://github.com/kubernetes-sigs/gateway-api/issues/1061)
- [L4 Gateway Sharing](https://github.com/kubernetes-sigs/gateway-api/issues/1062)

At this point, we believe that all of the above can be added in the future with [backwards compatible changes/additions](https://docs.google.com/document/d/1NN-IDoN-mIxtv6f7JR_G9kK1kKbZHMqQPcUSy6tPDbA/edit). Note: this is not saying we should or will make these additions, just that they will still be possible after graduating to beta.

### Goals of this review

Since we're an official Kubernetes API, we need formal approval for this change from SIG-Network API Reviewers, assigned below. Of course feedback from others is also welcome here. Our goal is to complete this process by mid to late April.

Note: The aim is not to merge this PR, just to provide a way to review what has changed between v0.4 and v0.5.

/assign @danwinship @khenidak @thockin

### CODE REVIEW

File: apis/v1alpha2/httproute_types.go [hunk 1]
```diff
@@ -704,6 +769,31 @@ type HTTPRequestRedirectFilter struct {
...
+// Support: Extended
+type HTTPURLRewriteFilter struct {
+	// Hostname is the value to be used to replace the Host header value during
+	// forwarding.
+	//
+	// Support: Extended
+	//
+	// <gateway:experimental>
+	// +optional
+	Hostname *Hostname `json:"hostname,omitempty"`
```
Comment (robscott) [inline apis/v1alpha2/httproute_types.go:786]:
```suggestion
	Hostname *PreciseHostname `json:"hostname,omitempty"`
```

File: apis/v1alpha2/shared_types.go [hunk 2]
```diff
@@ -401,3 +430,86 @@ type AnnotationKey string
...
 type AnnotationValue string
+
+// AddressRouteMatches defines AddressMatch rules for inbound traffic according to
+// source and/or destination address of a packet or connection.
+type AddressRouteMatches struct {
+	// SourceAddresses indicates the originating (source) network
+	// addresses which are valid for routing traffic.
+	//
+	// Support: Extended
+	SourceAddresses []AddressMatch `json:"sourceAddresses"`
```
Comment (thockin) [inline apis/v1alpha2/shared_types.go:441]: I'm suspicious of the need for this.  It feels like firewall behavior, but it doesn't purport to drop non-matching traffic, just to ignore it, IIUC?  That seems potentially problematic for some impls (e.g. iptables would not match and allow the packet to continue for further processing).

Can you help me understand what it is for?

File: apis/v1alpha2/shared_types.go [hunk 3]
```diff
@@ -401,3 +430,86 @@ type AnnotationKey string
...
+
+	// DestinationAddresses indicates the destination network addresses
+	// which are valid for routing traffic.
+	//
+	// Support: Extended
+	DestinationAddresses []AddressMatch `json:"destinationAddresses"`
```
Comment (thockin) [inline apis/v1alpha2/shared_types.go:447]: How is this different than listener?  Is this related to SNI?  No?

I'd like to understand how this is used - hostname isn't part of a TCP or UDP packet.  Is someone planning to do DNS lookups and filter on that?

File: apis/v1alpha2/shared_types.go [hunk 4]
```diff
@@ -401,3 +430,86 @@ type AnnotationKey string
...
+	// must respect the connection metadata from proxy-protocol
+	// in the match logic implemented for these address values.
+	//
+	// Examples: `1.2.3.4`, `128::1`, `my-named-address`.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Value string `json:"value"`
```
Comment (thockin) [inline apis/v1alpha2/shared_types.go:482]: IPv4, IPv6, hostname are unambiguously parseable (with some effort) - is this justified?

If we want to keep this, should it follow the same 1-field-per enum value pattern?

File: apis/v1alpha2/shared_types.go (hunk 3 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:447]: I think there are at least two key use cases here:

1. For routing requests going through Gateway(s) that are listening on a wide set of addresses
2. For routing requests not going through Gateway(s) such as mesh implementations

I think @shaneutt and/or @howardjohn might be able to add some more detail for potential use cases here.

File: apis/v1alpha2/shared_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:441]: Good question, it's been a while since we reviewed this, and I can't remember the specific use cases for source IP matching, maybe @shaneutt does?

File: apis/v1alpha2/shared_types.go (hunk 4 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:482]: I think it's at least possible that we'd add additional types in the future that may not be as easy to differentiate between. Even with named address, it's possible that in some systems, a name is just a number, or a name is a valid hostname, or something else. I could get on board with the 1-field-per enum pattern here though.

File: apis/v1alpha2/shared_types.go (hunk 3 above)
Comment (howardjohn) [reply apis/v1alpha2/shared_types.go:447]: The strongest use case, IMO, is for east-west traffic that can convey a destination address. This could be from CONNECT, PROXY protocol, service mesh redirection, etc.

For example, if we want to say "All requests to httpbin should use canary logic":

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: httpbin-canary
spec:
  parentRefs:
  - name: my-gateway
  matches:
  - destinationAddresses:
      type: NamedAddress
      value: httpbin.default.svc.cluster.local
  rules:
  - backendRefs:
    - name: httpbin-v2
      port: 80
      weight: 1
    - name: httpbin-v1
      port: 80
      weight: 1
```

(note: HTTP doesn't actually have this field currently, but the same applies to all protocols).

Matching only on Host header is not actually sufficient, IMO. For example, consider if we are trying to configure sending traffic to an nginx proxy, which then in turn does Host header routing (that is, the application does `curl nginx.default.svc.cluster.local -H "Host: host-for-nginx"`. Matching `hostnames: [nginx.default.svc.cluster.local]` would not help here.

**Comment by danwinship** [comment]
> The proposed API is largely identical to what we released in v1alpha2, with the following exceptions:
> 
> 1. We want to graduate GatewayClass, Gateway, and HTTPRoute to v1beta1 ([More Details](https://docs.google.com/document/d/1F3dcs_J8Y6GVlfMGjRtom5ljovnKa_MMIC34GJxUdBs/edit))

I see no references to "v1beta1" here... I had assumed when I read this that you'd be duplicating the "graduating" parts of `apis/v1alpha2/` into a new `apis/v1beta1/`, so that people using only beta APIs could do

    apiVersion: gateway.networking.k8s.io/v1beta1
    kind: Gateway
    ...

**Comment by youngnick** [comment]
@danwinship I think that's the plan for actually cutting the release, but I think that @robscott felt this review would be easier if you didn't need to review files that were just all added lines (since a `v1beta1` would be net-new and not produce a useful diff.)

The plan is absolutely to take the three resources mentioned and copy them into a `v1beta1` version however.

**Comment by robscott** [comment]
@danwinship Good point, I can make a separate PR with `/apis/v1beta1`, I was just trying to make it as easy as possible to see what was actually changing. The new v1beta1 types will be identical to the v1alpha2 types shown here, so I'd hoped just seeing a diff covering the current state with what you'd currently reviewed would be more manageable. Let me know if adding `/apis/v1beta1` at this point would be helpful.

File: apis/v1alpha2/httproute_types.go [hunk 5]
```diff
@@ -664,7 +684,42 @@ type HTTPRequestHeaderFilter struct {
...
+	Type HTTPPathModifierType `json:"type"`
+
+	// Substitution defines the HTTP path value to substitute. An empty value
+	// ("") indicates that the portion of the path to be changed should be
+	// removed from the resulting path. For example, a request to "/foo/bar"
+	// with a prefix match of "/foo" would be modified to "/bar".
+	//
+	// <gateway:experimental>
+	// +kubebuilder:validation:MaxLength=1024
+	Substitution string `json:"substitution"`
```
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:718]: Will be fixed in #1124

File: apis/v1alpha2/shared_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:441]: Discussed in community meeting and it seems like we may need to rethink this a bit, more discussion in #1127.

File: apis/v1alpha2/shared_types.go (hunk 3 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:447]: Discussed in community meeting and it seems like we may need to rethink this a bit, more discussion in #1127.

**Comment by youngnick** [comment]
Looks like this got closed in error?

File: apis/v1alpha2/shared_types.go [hunk 6]
```diff
@@ -253,12 +284,10 @@ type RouteStatus struct {
...
 // PreciseHostname is the fully qualified domain name of a network host. This matches
-// the RFC 1123 definition of a hostname with 2 notable exceptions:
-//
-// 1. IPs are not allowed.
-// 2. A hostname may not be prefixed with a wildcard label (`*.`).
+// the RFC 1123 definition of a hostname with 1 notable exception that
+// numeric IP addresses are not allowed.
 //
-// Hostname can be "precise" which is a domain name without the terminating
+// PreciseHostname can be "precise" which is a domain name without the terminating
```
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/shared_types.go:290]: Will be fixed in https://github.com/kubernetes-sigs/gateway-api/pull/1124

File: apis/v1alpha2/httproute_types.go [hunk 7]
```diff
@@ -227,11 +227,10 @@ const (
 	// Matches based on a URL path prefix split by `/`. Matching is
 	// case sensitive and done on a path element by element basis. A
 	// path element refers to the list of labels in the path split by
-	// the `/` separator. A request is a match for path _p_ if every
-	// _p_ is an element-wise prefix of the request path.
+	// the `/` separator. When specified, a trailing `/` is ignored.
 	//
-	// For example, `/abc`, `/abc/` and `/abc/def` match the prefix
-	// `/abc`, but `/abcd` does not.
+	// For example. the paths `/abc`, `/abc/`, and `/abc/def` would all match
```
Comment (candita) [inline apis/v1alpha2/httproute_types.go:232]: nit
```suggestion
	// For example, the paths `/abc`, `/abc/`, and `/abc/def` would all match
```

**Comment by danwinship** [comment]
/unassign
(I skimmed over it, but I haven't used this API at all, or thought about it since the last review, so I don't know the context of most of these changes, etc. None of it seems obviously wrong...)

File: apis/v1alpha2/httproute_types.go [hunk 8]
```diff
@@ -664,7 +684,42 @@ type HTTPRequestHeaderFilter struct {
...
+	// removed from the resulting path. For example, a request to "/foo/bar"
+	// with a prefix match of "/foo" would be modified to "/bar".
+	//
+	// <gateway:experimental>
+	// +kubebuilder:validation:MaxLength=1024
+	Substitution string `json:"substitution"`
+}
+
+// HTTPRequestRedirect defines a filter that redirects a request. This filter
+// MUST not be used on the same Route rule as a HTTPURLRewrite filter.
```
Comment (candita) [inline apis/v1alpha2/httproute_types.go:722]: nit
```suggestion
// MUST NOT be used on the same Route rule as a HTTPURLRewrite filter.
```

File: apis/v1alpha2/httproute_types.go [hunk 9]
```diff
@@ -704,6 +769,31 @@ type HTTPRequestRedirectFilter struct {
 	StatusCode *int `json:"statusCode,omitempty"`
 }
 
+// HTTPURLRewriteFilter defines a filter that modifies a request during
+// forwarding. At most one of these filters may be used on a Route rule. This
+// may not be used on the same Route rule as a HTTPRequestRedirect filter.
```
Comment (candita) [inline apis/v1alpha2/httproute_types.go:774]: nit - for consistency
```suggestion
// MUST NOT be used on the same Route rule as a HTTPRequestRedirect filter.
```

File: apis/v1alpha2/shared_types.go [hunk 10]
```diff
@@ -84,6 +86,35 @@ type ParentRef struct {
 	//
 	// +optional
 	SectionName *SectionName `json:"sectionName,omitempty"`
+
+	// Port is the network port this Route targets. It can be interpreted
+	// differently based on the type of parent resource:
+	//
+	// * Gateway: All listeners listening on the specified port that also
```
Comment (candita) [inline apis/v1alpha2/shared_types.go:93]:
```suggestion
	// If the parent resource is a Gateway, all listeners listening on the specified port that also
```

File: apis/v1alpha2/shared_types.go [hunk 11]
```diff
@@ -84,6 +86,35 @@ type ParentRef struct {
...
+	// support this kind of Route(and select this Route). It's not
```
Comment (candita) [inline apis/v1alpha2/shared_types.go:94]: This sentence is a little unclear.  Is it a definition of the Gateway interpretation of Port?  Or is Gateway a type of parent resource?

File: apis/v1alpha2/shared_types.go [hunk 12]
```diff
@@ -401,3 +430,86 @@ type AnnotationKey string
...
+	// implementation to handle (and add validation for) according to their
+	// own needs.
+	//
+	// For IPAddress the implementor may expect either IPv4 or IPv6.
+	//
+	// Support: Core (IPAddress)
+	// Support: Custom (NamedAddress)
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=IPAddress;NamedAddress
```
Comment (candita) [inline apis/v1alpha2/shared_types.go:464]: The Enum lacks Hostname
```suggestion
	// +kubebuilder:validation:Enum=IPAddress;NamedAddress;Hostname
```

**Comment by candita** [comment]
@robscott looks good to me, as someone new to the review.  I just had a few nits and minor comments.

File: apis/v1alpha2/shared_types.go (hunk 3 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: This has been removed from the API while we reconsider how it could work.

File: apis/v1alpha2/shared_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: This has been removed from the API while we reconsider how it could work.

File: apis/v1alpha2/shared_types.go (hunk 12 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: This ended up getting pulled out entirely with the latest update.

File: apis/v1alpha2/shared_types.go (hunk 11 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:90]: Good catch - will be fixed by #1160

File: apis/v1alpha2/httproute_types.go (hunk 9 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:789]: Good catch - will be fixed by #1160

File: apis/v1alpha2/httproute_types.go (hunk 8 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:737]: Good catch - will be fixed by #1160

File: apis/v1alpha2/httproute_types.go [hunk 13]
```diff
@@ -664,7 +699,42 @@ type HTTPRequestHeaderFilter struct {
...
+
+const (
+	// This type of modifier indicates that the complete path will be replaced
+	// by the path redirect value.
+	AbsoluteHTTPPathModifier HTTPPathModifierType = "Absolute"
+
+	// This type of modifier indicates that any prefix path matches will be
+	// replaced by the substitution value. For example, a path with a prefix
+	// match of "/foo" and a ReplacePrefixMatch substitution of "/bar" will have
+	// the "/foo" prefix replaced with "/bar" in matching requests.
```
Comment (thockin) [inline apis/v1alpha2/httproute_types.go:713]: Is this a path-token match or a string match?  If the input path is `/foo/qux` it will become `/bar/qux`, yes?  What if the input was `/fooqux` ?

Comment (aojea) [reply apis/v1alpha2/httproute_types.go:713]: the ingress one is `path` token https://kubernetes.io/docs/concepts/services-networking/ingress/#path-types , should this be the same?

File: apis/v1alpha2/shared_types.go [hunk 14]
```diff
@@ -155,22 +182,59 @@ type BackendRef struct {
...
-	ConditionRouteAccepted RouteConditionType = "Accepted"
+	//
+	// Possible reasons for this condition to be true are:
+	//
+	// * "Accepted"
+	//
+	// Controllers may raise this condition with other reasons,
+	// but should prefer to use the reasons listed above to improve
+	// interoperability.
+	RouteConditionAccepted RouteConditionType = "Accepted"
```
Comment (aojea) [inline apis/v1alpha2/shared_types.go:199]: naive question, "Accepted" is a Type?

File: apis/v1alpha2/httproute_types.go (hunk 7 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Updated, thanks!

File: apis/v1alpha2/httproute_types.go (hunk 8 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Updated, thanks!

File: apis/v1alpha2/httproute_types.go (hunk 9 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Updated, thanks!

File: apis/v1alpha2/shared_types.go (hunk 14 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:204]: Not sure I understand the question. We're defining it as a type along with all the corresponding reasons we recommend setting for that condition. This matches the pattern we've used elsewhere for status conditions and reasons.

File: apis/v1alpha2/shared_types.go (hunk 11 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Updated, thanks!

File: apis/v1alpha2/shared_types.go (hunk 14 above)
Comment (aojea) [reply apis/v1alpha2/shared_types.go:204]: ah, I saw below reason "accepted" and type "ResolvedRefs" so I get confused, it was curiosity , disregard it

File: apis/v1alpha2/shared_types.go (hunk 4 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Good catch, this was not clearly documented in the context of rewrites and redirects, I've filed https://github.com/kubernetes-sigs/gateway-api/pull/1189 to fix that.

File: apis/v1alpha2/httproute_types.go (hunk 13 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:718]: Good catch, this was not clearly documented in the context of rewrites and redirects, I've filed https://github.com/kubernetes-sigs/gateway-api/pull/1189 to fix that.

**Review by thockin — COMMENTED** [review]
Discussed in real-time - couple small followups Rob noted.  LGTM.  Thanks.

**Comment by youngnick** [comment]
So, I think that once we resolve #1200, #1201, and #1202, we can close this out as complete?

**Comment by thockin** [comment]
SGTM

**Comment by youngnick** [comment]
Yes, I think we may just close this out - this is really a placeholder to give the API reviewers an easier time. Unless there's something else @robscott wants to do here?

**Comment by robscott** [comment]
Yep, I think this was good to go. All the changes reviewed here were released as part of v0.5.0-rc1.
